### ISSUE TITLE
Docker logs (Java stack trace that are long) NOT meaningful in Splunk

### ISSUE THREAD

**Issue by jambajaar** [rancher/rancher#10289]
**Problem**

_What?_
Docker log driver treats multi-line long strings (of a single logging event) from stack traces of Java applications as separate logging events. Basically, it splits a long string into multiple lines and adds some timestamp at the beginning of each new line.
 
@wjimenez5271 seems to have pursued this Docker.  There does not seem to be any fix from Docker other than workaround suggestions from the community
https://github.com/moby/moby/issues/22920

_Why this important to a devops org?_
This is important for orgs that have many Java-based apps pending containerization, where Splunk is their primary log/analytics tool.  When the logs are ported into Splunk, the docker logs of Java stack traces become hard to follow.  So, in a way, containerizing is a regression for the DevOps teams of such Java apps.  And, they may hold moving their apps to docker containers / Rancher.

**Recommended workaround from Rancher, in the absence of a product fix:**

We could use Fluentd to parse the multi-line log to one line, like stack traces, and ship to Splunk.

Steps:

1. Get the log path, if the log uses stdout the log path in the host is /var/lib/docker/containers/, if the log file is defined by user, could mount the path to the host or other way that could let Fluentd access

2. Install Fluentd and plugin.  Example:
2.1 Ubuntu and Ruby
```
sudo apt install ruby  # install ruby, better version 2.3
gem install fluentd -v "~> 0.12.0" --no-ri --no-rdoc     # latest one is 0.14, 0.12 is stable version
gem install fluent-plugin-splunk-http-eventcollector --no-ri --no-rdoc
gem install fluent-plugin-detect-exceptions --no-ri --no-rdoc
```

More install way could be reached here https://docs.fluentd.org/v0.12/categories/installation

3.  Configure Fluentd to get container log, and use plugin to format the log from multi-line to one line.  For example, the Fluentd configure file like below. It includes two plugins, we use gem install to install them above.  https://github.com/brycied00d/fluent-plugin-splunk-http-eventcollector  Splunk output plugin for Fluent event collector.  https://github.com/GoogleCloudPlatform/fluent-plugin-detect-exceptions  A plugin scans a log stream text messages or JSON records for multi-line exception stack traces
 
Fluentd configure file, more details could be seen in the below.

![image001](https://user-images.githubusercontent.com/11086008/32532552-05af0020-c400-11e7-8af5-3506fc968e71.png)

4. The output log in Splunk will look like below.

![image002](https://user-images.githubusercontent.com/11086008/32532577-2898e7a4-c400-11e7-9d52-c5bacc15f7f5.png)

**Comment by aiwantaozi** [comment]
Several ways to ship log to fluentd:
* Install fluend in the host as above
* Run fluentd container and mount the log path to it
* Use fluentd log driver directly. Attention: in this way, the docker logs command is not available for this logging driver.  https://docs.docker.com/engine/admin/logging/fluentd/

#### More details about the Fluentd config:

```
<source>
  type tail
  # path for the log file, * could match all
  path /var/lib/docker/containers/*/*.log
  pos_file /fluentd/etc/fluentd-containers.log.pos
  time_format %Y-%m-%dT%H:%M:%S.%NZ
  tag prefix.myparse.*
  format json
  read_from_head true
</source>

<match prefix.**>
  @type detect_exceptions 

  # remove_tag_prefix remove from the input tag when outputting a record. 
  # required now, not configure this may cause infinite loop
  remove_tag_prefix prefix 

  # Name of the field in the JSON record that contains the single-line log messages
  message log

  # Supported values: java, js, csharp, python, go, ruby, php, all. Default: all.
  languages java, python

  # Interval in seconds after which a (possibly not yet complete) buffered exception stack shall be forwarded. If not set, incomplete exceptions stacks are not flushed.
  multiline_flush_interval 0.1
</match>

<match myparse.**> # attention: about match have a field remove_tag_prefix prefix, do the tag will change from prefix.myparse.*** to myparse.**
  @type splunk-http-eventcollector

  # server: Splunk server host and port
  server localhost:8088
  # all_items config use the original record, not only the message field in the original record
  all_items true

  # protocol: Connect to Splunk server via 'http' or 'https'
  # default: https
  protocol http

  # verify: SSL server verification
  # default: true
  # verify false

  # token: hec the token issued
  # How to get a token: http://dev.splunk.com/view/event-collector/SP-CAAAE7C
  token test_token

  #
  # Event Parameters
  #

  # host: 'host' parameter passed to Splunk
  host testhost

  # sourcetype: 'sourcetype' parameter passed to Splunk
  # default: fluent
  sourcetype testsource

  #
  # Formatting Parameters
  #

  # time_format: the time format of each event
  # value: none, unixtime, localtime, or any time format string
  # default: localtime
  time_format localtime

  format json

  #
  # Buffering Parameters
  #

  # Standard parameters for buffering.  See documentation for details:
  #   http://docs.fluentd.org/articles/buffer-plugin-overview
  buffer_type memory
  buffer_queue_limit 16

  # buffer_chunk_limit: The maxium size of POST data in a single API call.
  # This value should be reasonablly small since the current implementation
  # of out_splunk-http-eventcollector converts a chunk to POST data on memory before API calls.
  # The default value should be good enough.
  buffer_chunk_limit 8m

  # flush_interval: The interval of API requests.
  # 
  # Make sure that this value is sufficiently large to make successive API calls.
  # Note that a different 'source' creates a different API POST, each of which may
  # take two or more seconds.  If you include "{TAG}" in the source parameter and
  # this 'match' section recieves many tags, a single flush may take long time.
  flush_interval 1s
</match>
```

**Comment by nsaud01** [comment]
Does fluent-plugin-detect-exceptions work as a plugin on its own? I keep getting this error when running my container:

````
2017-11-14 20:19:26 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/fluent.conf"
/usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require': cannot load such file -- /usr/lib/ruby/gems/2.3.0/gems/fluent-plugin-detect-exceptions-0.0.8/lib/fluent/plugin/out_detect_exceptions (LoadError)
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/registry.rb:102:in `block in search'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/registry.rb:99:in `each'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/registry.rb:99:in `search'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/registry.rb:44:in `lookup'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/plugin.rb:146:in `new_impl'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/plugin.rb:104:in `new_output'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/agent.rb:128:in `add_match'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/agent.rb:72:in `block in configure'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/agent.rb:64:in `each'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/agent.rb:64:in `configure'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/root_agent.rb:112:in `configure'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/engine.rb:131:in `configure'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/engine.rb:96:in `run_configure'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/supervisor.rb:770:in `run_configure'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/supervisor.rb:553:in `dry_run'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/supervisor.rb:571:in `supervise'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/supervisor.rb:476:in `run_supervisor'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/lib/fluent/command/fluentd.rb:310:in `<top (required)>'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /usr/lib/ruby/2.3.0/rubygems/core_ext/kernel_require.rb:55:in `require'
	from /usr/lib/ruby/gems/2.3.0/gems/fluentd-0.14.22/bin/fluentd:5:in `<top (required)>'
	from /usr/bin/fluentd:22:in `load'
	from /usr/bin/fluentd:22:in `<main>'
````

**Comment by manuel220x** [comment]
@nsaud01 For the error that you described, seems like a problem with permissions on files under: `/usr/lib/ruby/gems/2.3.0/gems/fluent-plugin-detect-exceptions-0.0.8/lib/fluent/plugin/`

**Comment by loganhz** [comment]
Please let us know in a comment if you can still reproduce the issue, and we'll reopen it

**Comment by ifeldshteyn** [comment]
Hello, I am seeing a similar issue as the user before. Is this also some kind of permission issue and if it is what folder/files should I chown/chmod? 

2019-01-31 22:59:44 +0000 [info]: parsing config file is succeeded path="/fluentd/etc/fluent.conf"
Traceback (most recent call last):
        28: from /usr/bin/fluentd:23:in `<main>'
        27: from /usr/bin/fluentd:23:in `load'
        26: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/bin/fluentd:8:in `<top (required)>'
        25: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        24: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
        23: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/command/fluentd.rb:310:in `<top (required)>'
        22: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/supervisor.rb:502:in `run_supervisor'
        21: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/supervisor.rb:597:in `supervise'
        20: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/supervisor.rb:579:in `dry_run'
        19: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/supervisor.rb:795:in `run_configure'
        18: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/engine.rb:96:in `run_configure'
        17: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/engine.rb:131:in `configure'
        16: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/root_agent.rb:112:in `configure'
        15: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/agent.rb:64:in `configure'
        14: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/agent.rb:64:in `each'
        13: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/agent.rb:72:in `block in configure'
        12: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/agent.rb:128:in `add_match'
        11: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/plugin.rb:104:in `new_output'
        10: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/plugin.rb:146:in `new_impl'
         9: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/registry.rb:44:in `lookup'
         8: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/registry.rb:99:in `search'
         7: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/registry.rb:99:in `each'
         6: from /usr/lib/ruby/gems/2.5.0/gems/fluentd-1.3.2/lib/fluent/registry.rb:102:in `block in search'
         5: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         4: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:59:in `require'
         3: from /usr/lib/ruby/gems/2.5.0/gems/fluent-plugin-detect-exceptions-0.0.11/lib/fluent/plugin/out_detect_exceptions.rb:16:in `<top (required)>'
         2: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:39:in `require'
         1: from /usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:135:in `rescue in require'
/usr/lib/ruby/2.5.0/rubygems/core_ext/kernel_require.rb:135:in `require': cannot load such file -- fluent/plugin/exception_detector (LoadError)

**Comment by loganhz** [comment]
@ifeldshteyn Could you submit a new issue with reproduce steps, please? Thanks
